### PR TITLE
Vulkan rendering support on Android

### PR DESCRIPTION
**Pull request by pouleyKetchoupp**
This change enables Vulkan rendering for Android, based on the SurfaceView implementation from PR #36603. There are still some issues to solve, but as a first step Android is now functional again on the 4.0 branch.

Only Vulkan rendering is supported for now, but I've kept the old GL implementation along with this changes, so it should be easy to enable GLES2 again when it's ready.

CC @RandomShaper @m4gr3d @akien-mga @reduz 

**Remaining issues:**

- Missing 3D support (needs a new renderer implementation & shaders)
Shaders currently fail because of descriptor set limits (tested on Galaxy Note 5):
```
ERROR: On shader stage 'Vertex', uniform 'transforms' uses a set (4) index larger than what is supported by the hardware (4).
   at: shader_create (drivers\vulkan\rendering_device_vulkan.cpp:3728) - Condition "set >= limits.maxBoundDescriptorSets" is true. returned: RID()
```

- Colors are off, maybe a texture format issue?
Example with `TextureRect` :

| Normal | Android |
| ------------- | ------------- |
|   |   |

- Vulkan validation layers are not working (fails on `vkGetInstanceProcAddr`)
It's disabled on Android for now to avoid crashes.

- Errors when using format list in `RenderingDeviceVulkan::texture_create`
`vmaCreateImage` fails with `VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT` (returns `VK_ERROR_OUT_OF_HOST_MEMORY`)
It's disabled on Android for now to avoid crashes.

- Errors in `RenderingDeviceVulkan::compute_pipeline_create`
`vkCreateComputePipelines` returns `VK_ERROR_INITIALIZATION_FAILED` in some cases
```
ERROR: vkCreateComputePipelines returned error -3
   at: compute_pipeline_create (drivers\vulkan\rendering_device_vulkan.cpp:5143) - Condition "err" is true. returned: RID()
```

- Missing VR support
There's specific initialization for GL, based on `xrMode`, it's not handled for Vulkan.

- Missing fallback to GLES2

- Building templates with Android API 18 to 23
Vulkan support should be disabled in this case.

- ~Properly handle `onVkSurfaceChanged` event to update the rendering system~
~Right now, pausing and resuming the app causes it to restart completely instead.~
**edit**: *addressed in #39004*

**Implementation Notes:**

- New Android SDK requirements for building templates (Vulkan headers and libs):
NDK version: 20.0.5594570
Min Android API: 24

- The Vulkan loader is taken from the Android NDK, because the one we're using for other platforms doesn't support Android platform.
https://github.com/KhronosGroup/Vulkan-Loader/issues/96

- `VkRenderer` calls GodotLib directly, so I got rid of `vk_renderer_jni`. It seems easier to have one common place for calls from GL or Vulkan implementations.

- Fix for `GodotInstrumentation` in `AndroidManifest.xml`
This is used to force the app to restart.
I've kept the same hack that was used for GL to restart the whole app after paused, but we might want to improve it later by restarting only the rendering system.

- Plugins:
Added vulkan specific callbacks in addition to the GL ones, but maybe the interface could be harmonized between the two to make it easier for the user.

*Bugsquad edit:* Fixes #36919.

### CODE REVIEW

File: drivers/vulkan/SCsub
Comment (akien-mga) [inline drivers/vulkan/SCsub]: How presumptuous ;)
This should be `elif env['platform'] == "android":`

File: drivers/vulkan/SCsub
Comment (akien-mga) [inline drivers/vulkan/SCsub]: When false, this basically disables the whole `if env['builtin_vulkan']`, i.e. it's not using the builtin vulkan code. So instead of doing it this way, simply do:
```
if env['builtin_vulkan'] and not env['platform'] == "android":
...
elif env['platform'] == "android":
...
```

File: drivers/vulkan/SCsub
Comment (akien-mga) [reply drivers/vulkan/SCsub]: Actually, see above comment.

File: platform/android/java/lib/src/org/godotengine/godot/xr/regular/RegularContextFactory.java
Comment (akien-mga) [inline platform/android/java/lib/src/org/godotengine/godot/xr/regular/RegularContextFactory.java]: Is this really solved though?

File: platform/android/vulkan_context_android.cpp
Comment (akien-mga) [inline platform/android/vulkan_context_android.cpp]: Trailing `*/` misaligned.

File: platform/android/vulkan_context_android.h
Comment (akien-mga) [inline platform/android/vulkan_context_android.h]: Misaligned.

File: platform/android/java/lib/src/org/godotengine/godot/xr/regular/RegularContextFactory.java
Comment (pouleyKetchoupp) [reply platform/android/java/lib/src/org/godotengine/godot/xr/regular/RegularContextFactory.java]: I figured handling Vulkan wasn't going to take place in this file since it's GL specific, but I can put the comment back if you want.

**Comment by pouleyKetchoupp** [comment]
@akien-mga Thanks for the review! I've just pushed some changes based on your comments.

File: platform/android/detect.py
Comment (akien-mga) [inline platform/android/detect.py]: I guess Vulkan support should be enabled conditionally if `get_platform(env["ndk_platform"]) >= 24`. Unless we decide otherwise, it should remain possible for now to compile custom Android templates with only OpenGL ES.

API 18 to 23 are currently still close to 40% of the market share: https://developer.android.com/about/dashboards

Comment (akien-mga) [reply platform/android/detect.py]: Note that if more work is needed to preserve the possibility to build against API 18-23, it can be left for a later PR.

**Review by akien-mga — APPROVED** [review]
Buildsystem and C++ changes look good. I'll let @m4gr3d review the Java/Kotlin parts more in-depth.

File: platform/android/detect.py
Comment (pouleyKetchoupp) [reply platform/android/detect.py]: That's what I had in mind. I've updated the PR description to reflect this in the todo list.

**Comment by reduz** [comment]
I think this is OK, although it may need to be refactored to use the DisplayServer code that will be merged in a week or two.

File: platform/android/java/lib/src/org/godotengine/godot/plugin/GodotPlugin.java
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/plugin/GodotPlugin.java]: I'd merge these with the `GL*` callbacks by renaming them to make them generic. For example, `onRenderDrawFrame`, `onRenderSurfaceChanged`, `onRenderSurfaceCreated`.

File: platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt]: You can use `val` here since the variable is `final`.

File: platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt]: Since you removed the `native*` callbacks, how are they propagated to the native layer?

File: platform/android/java/lib/src/org/godotengine/godot/GodotLib.java
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/GodotLib.java]: The method name should be updated to be camelCase (`surfaceCreated`).

In the native layer, the logic for this method is almost identical to `newcontext`. The two should be merged into one.

File: platform/android/vulkan_context_android.h
Comment (m4gr3d) [inline platform/android/vulkan_context_android.h]: It's likely that more vulkan logic will be added inside and outside of this class. Can you restore the `vulkan` directory and keep all android platform vulkan specific logic (including this class) in there.

**Review by m4gr3d — CHANGES_REQUESTED** [review]
@pouleyKetchoupp Nice job! I've added some feedback to be addressed.

File: platform/android/java/lib/src/org/godotengine/godot/GodotViewVulkan.java
Comment (pouleyKetchoupp) [reply platform/android/java/lib/src/org/godotengine/godot/GodotViewVulkan.java]: No I don't imagine there would be more classes, unless we need a specific one to handle `xrMode` (I haven't looked at that part of the code at all).

The idea was just to keep Vulkan-specific logic in `VkSurfaceView`, and implement the common interface with GL in `GodotViewVulkan`, but instead I can make `VkSurfaceView` implement `GodotViewInterface` directly if that makes more sense to you.

File: platform/android/java/lib/src/org/godotengine/godot/plugin/GodotPlugin.java
Comment (pouleyKetchoupp) [reply platform/android/java/lib/src/org/godotengine/godot/plugin/GodotPlugin.java]: Do you mean that the common functions would take both a `Surface` and a `GL10` as arguments, and one of them would be null depending on the renderer used?

Or maybe we could simply pass a `SurfaceView` argument instead, but I don't have an example of these callbacks being used so I'm not sure what the actual use cases are.

File: platform/android/java/lib/src/org/godotengine/godot/Godot.java
Comment (pouleyKetchoupp) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java]: Yes that's the reason.

I can do it by command line instead as you suggested, but in this case we'll have less options when we want to handle the fallback to OpenGL when Vulkan is not supported on the device (at this point I'm not sure if we'll do that in c++ or java).

This part of the code is supposed to handle the case where `Main::setup` is called on the UI thread and it seems to work fine so far:
https://github.com/godotengine/godot/blob/989a223c5a6d6ba5b1b098be8983888cb49b2525/platform/android/java_godot_lib_jni.cpp#L197-L199

What part of the setup do you think is going to cause problems when not run on the render thread?

File: platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt
Comment (pouleyKetchoupp) [reply platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt]: They are propagated through calls to `GodotLib` from `VkRenderer`, like `GodotRenderer` does for OpenGL.

File: platform/android/java/lib/src/org/godotengine/godot/GodotViewVulkan.java
Comment (m4gr3d) [reply platform/android/java/lib/src/org/godotengine/godot/GodotViewVulkan.java]: Good point about `xrMode` possibly needing another implementation, and also about the separation of concern.
I agree with your current implementation then; let's keep them separate for now especially since we're still early in the vulkan integration process.
If we find out once the integration is complete that we don't need them separate, then we can merge them together.

File: platform/android/java/lib/src/org/godotengine/godot/Godot.java
Comment (m4gr3d) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java]: My main worry is regarding thread safety since `setup` initializes some data structures that `setup2` and the rest of the logic access at runtime. 
With your changes, `setup` concludes before the render thread is started, so data access should be safe, especially given that Godot's data structures should be thread-safe.

@akien-mga @reduz It would be good to have someone familiar with Godot's C/C++ thread safety validates these changes as well.

**Comment by pouleyKetchoupp** [comment]
Thanks @m4gr3d! I've just pushed some changes that should address your previous comments.

So here are the remaining things to do before this PR can be merged:
- Validate it's safe to call `Main::setup()` on the UI thread, then `Main::setup2(Thread::get_caller_id());` on the render thread after it's created
- Change the calls for plugins after PR #37174 is merged
- Refactor the c++ part after the DisplayServer branch is merged

**Review by m4gr3d — APPROVED** [review]
Looks great!

One feedback to address, alongside your comment.

File: platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt
Comment (pouleyKetchoupp) [reply platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt]: Ah yes, good point! Right now, I'm not handling the surface re-creation properly, so the workaround is to reset the app by calling `newcontext`. Instead it should update the surface in the native code, but I haven't looked into that yet.

The consequence is the app gets restarted when paused and resumed again.

It's important to fix and I'll see if I have time for it in the coming days, but I don't think it's a blocker to merge this PR. In the meantime I'll add a comment in the code and in the PR description.

Comment (m4gr3d) [reply platform/android/java/lib/src/org/godotengine/godot/vulkan/VkRenderer.kt]: Sounds good. 
For the moment, you could then remove the call to `newcontext` from `onVkSurfaceCreated` since it's always followed by `onVkSurfaceChanged` and leave a comment like you mentioned to restore the proper logic flow when surface recreation is properly handled.

File: platform/android/display_server_android.h
Comment (m4gr3d) [inline platform/android/display_server_android.h]: @pouleyKetchoupp Feel free to deprecate these methods (and in `GodotIO` as well). Video playback functionality will be offered in the future via an Android plugin instead of being built-in.

@akien-mga @reduz Do you know if this functionality is actively being used? If it's not, they can be removed from the `DisplayServer` class as well since base on the comment, the methods are only available on mobile.

Comment (pouleyKetchoupp) [reply platform/android/display_server_android.h]: Ok, I'm going to remove it then. From what I've seen, it's not used in the engine and the play method is not working at the moment anyway:
https://github.com/godotengine/godot/blob/0ecca1abd0716a5f876bed619b6abbbbf1d3f269/platform/android/java_godot_io_wrapper.cpp#L170-L172

Comment (pouleyKetchoupp) [reply platform/android/display_server_android.h]: I'm removing native video support on Android, but it might still be used for iOS though.

**Comment by pouleyKetchoupp** [comment]
@akien-mga @reduz I'm done with making the Android port compatible with display servers. So if these changes are ok, I think it might be ready to merge when this last thing is validated:
- `Main::setup()` is called on the UI thread, then `Main::setup2(Thread::get_caller_id());` on the render thread after it's created by the android app. The parameter passed in `setup2` allows to override `Thread::_main_thread_id`. It looks like it was made before for android but wasn't used recently. It seems to work so far but do you think it's safe to use it?

File: platform/android/display_server_android.cpp
Comment (reduz) [inline platform/android/display_server_android.cpp:64]: Android does support native video, you should be able to port it from 3.x branch

**Review by reduz — CHANGES_REQUESTED** [review]

File: platform/android/java/lib/src/org/godotengine/godot/Godot.java
Comment (reduz) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java:221]: Ah nevermind, this will probably need to be changed and probably done again in setup2.
https://github.com/godotengine/godot/blob/master/main/main.cpp#L404

File: platform/android/display_server_android.cpp
Comment (m4gr3d) [reply platform/android/display_server_android.cpp:64]: @reduz The 3.x implementation is brittle and lack flexibility. The thinking for 4.0 (and 3.2.2) is to provide video support via the revamped Android plugin system instead.

File: platform/android/java/lib/src/org/godotengine/godot/Godot.java
Comment (m4gr3d) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java:221]: > Ah nevermind, this will probably need to be changed and probably done again in setup2.
> https://github.com/godotengine/godot/blob/master/main/main.cpp#L404

This is already done in https://github.com/godotengine/godot/blob/master/platform/android/java_godot_lib_jni.cpp#L201.

Comment (m4gr3d) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java:221]: > I think there should not be a problem calling setup2() on a different thread as setup(), if setup2() will later become the rendering thread. That said, eventually we need to look at some point at optionally supporting a dedicated rendering thread on Android, given this works on Godot already in the rest of the platforms.

@reduz The thread that `setup2()` is invoked on is the dedicated rendering thread. Or were you referring to something else?

**Review by m4gr3d — APPROVED** [review]

**Comment by pouleyKetchoupp** [comment]
Thanks for the review @reduz! I've just added the #warning you asked for.

If everything's good around the display server code, this PR should be ready to merge.

**Comment by thebestnom** [comment]
after trying to rebase android editor onto this branch, I found some small things that I think should be refactored...
by my understanding and how it looks display server should process everything that the user input and outputs, that means that process_{input} (process_touch for example) should be moved to display_server_android from os_android as well, I can see that when I tried to merge my ( @m4gr3d for now, later I'll try your implementation) that things doesn't go where they should be

**Comment by pouleyKetchoupp** [comment]
@thebestnom Ok, that makes sense. Thanks for the feedback! I can change it some time this week, but if you decide to do it in the meantime, please let me know and feel free to open a PR :)

**Comment by thebestnom** [comment]
@pouleyKetchoupp working on it :)

**Comment by thebestnom** [comment]
@pouleyKetchoupp hey, I commited everything but it doesn't let me open PR into https://github.com/nekomatata/godot/tree/android-vulkan-rendering

**Comment by pouleyKetchoupp** [comment]
@thebestnom It looks like you would have to actually fork from my repo to be able to do that :/
I'll just cherry-pick the changes from your commit instead.

**Comment by thebestnom** [comment]
> It looks like you would have to actually fork from my repo to be able to do that :/

I'm not sure, I can open PR to different forks, I think the problem is that it's team only repo or somthing like that, but it doesn't matter that much...

I wanted to make some commant on the code in the PR... so Ill just say it here,
while I moved most of the things as is, I changed `process_event` to be `process_key_event` and moved the implementation from `java_godot_ lib_jni` to the display server cause it made no sense for it to be there

**Comment by pouleyKetchoupp** [comment]
Ok! I've seen the change with key events and it makes sense too. I've just pushed a new version that includes your changes.

File: platform/android/java/lib/src/org/godotengine/godot/Godot.java
Comment (reduz) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java:221]: oh no, it's the main thread for Godot. On android there is no render thread support, so you should assume both render and main should be the same for now.

I'm sure it could eventually be done, but I would get everything to work first.

**Comment by akien-mga** [comment]
Some comments from IRC:
```
23:39 <reduz> m4gr3d[m]: I think it's ok, just to make clear, Main::setup2() , Main::iteration() and the events that come from DisplayServer, should all be in the same thread
23:40 <reduz> m4gr3d[m]: which on android so far is the same thread used by Vulkan/OpenGL to render
23:40 <reduz> m4gr3d[m]: we could move the game logic to an other thread and leave the render thread only for rendering, but that will need some small changes to the engine, and I would probably prefer to implement this once everything else is working
```

Let's merge as is and refine in future PRs :)

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Tested a local build, it's not finding `vk_mem_alloc.h`. `#thirdparty/vulkan/` should be added to the `CPPPATH`.

**Comment by akien-mga** [comment]
> Tested a local build, it's not finding `vk_mem_alloc.h`. `#thirdparty/vulkan/` should be added to the `CPPPATH`.

It works after upgrading the NDK from `r20` to `r21` as `r21/sources/third_party/vulkan/src/layers/vk_mem_alloc.h` as been added, so we'll have to document r21 as minimum required version.

**Comment by m4gr3d** [comment]
> > Tested a local build, it's not finding `vk_mem_alloc.h`. `#thirdparty/vulkan/` should be added to the `CPPPATH`.
> 
> It works after upgrading the NDK from `r20` to `r21` as `r21/sources/third_party/vulkan/src/layers/vk_mem_alloc.h` as been added, so we'll have to document r21 as minimum required version.

@akien-mga @pouleyKetchoupp We can specify an NDK version in the `build.gradle` file as shown [here](https://developer.android.com/studio/projects/install-ndk#specific-version). 
We could use that to enforce the NDK version that's used.

**Comment by pouleyKetchoupp** [comment]
@m4gr3d In order to be more flexible when GLES2 is supported again, we could detect the NDK version during the build, and when the requirement is not met, show a warning and disable Vulkan support.

**Comment by m4gr3d** [comment]
> @m4gr3d In order to be more flexible when GLES2 is supported again, we could detect the NDK version during the build, and when the requirement is not met, show a warning and disable Vulkan support.

@pouleyKetchoupp that sounds overly complex and I'm not sure what the benefit would be done resolving the correct ndk version is a fairly simple task.

**Comment by thebestnom** [comment]
can't you just use `ndkVersion` in build.gradle? there is no benefit from using older ndk from my understanding

**Comment by Anutrix** [comment]
Can someone make a tracker issue for all the issues mentioned in first comment?

**Comment by pouleyKetchoupp** [comment]
@Anutrix Ok, I can make one.

**Comment by pouleyKetchoupp** [comment]
@Anutrix Tracker issue: #39033

**Pull request merged**
